pkg/kds/mux/client.go

```diff
@@ -284,79 +242,71 @@ func (c *client) startClusters(
 	errorCh chan error,
 ) {
 	client := mesh_proto.NewGlobalKDSServiceClient(conn)
 	log = log.WithValues("rpc", "healthcheck")
 	log.Info("starting")
 
-	go func() {
-		prevInterval := 5 * time.Minute
-		ticker := time.NewTicker(prevInterval)
-		defer ticker.Stop()
-		for {
-			log.Info("sending health check")
-			resp, err := client.HealthCheck(ctx, &mesh_proto.ZoneHealthCheckRequest{})
-			if err != nil && !errors.Is(err, context.Canceled) {
-				if status.Code(err) == codes.Unimplemented {
-					log.Info("health check unimplemented in server, stopping")
-					return
-				}
-				log.Error(err, "health check failed")
-				errorCh <- errors.Wrap(err, "zone health check request failed")
-			} else if interval := resp.Interval.AsDuration(); interval > 0 {
-				if prevInterval != interval {
-					prevInterval = interval
-					log.Info("Global CP requested new healthcheck interval", "interval", interval)
-				}
-				ticker.Reset(interval)
-			}
-
-			select {
-			case <-ticker.C:
-				continue
-			case <-stop:
-				log.Info("stopping")
+	prevInterval := 5 * time.Minute
+	ticker := time.NewTicker(prevInterval)
+	defer ticker.Stop()
+	for {
+		log.Info("sending health check")
+		resp, err := client.HealthCheck(ctx, &mesh_proto.ZoneHealthCheckRequest{})
+		if err != nil && !errors.Is(err, context.Canceled) {
+			if status.Code(err) == codes.Unimplemented {
+				log.Info("health check unimplemented in server, stopping")
 				return
 			}
+			log.Error(err, "health check failed")
+			errorCh <- errors.Wrap(err, "zone health check request failed")
+		} else if interval := resp.Interval.AsDuration(); interval > 0 {
+			if prevInterval != interval {
+				prevInterval = interval
+				log.Info("Global CP requested new healthcheck interval", "interval", interval)
+			}
+			ticker.Reset(interval)
 		}
-	}()
+
+		select {
+		case <-ticker.C:
+			continue
+		case <-ctx.Done():
+			log.Info("stopping")
+			return
+		}
+	}
 }
 
 func (c *client) handleProcessingErrors(
 	stream grpc.ClientStream,
 	log logr.Logger,
-	stop <-chan struct{},
 	processingErrorsCh chan error,
 	errorCh chan error,
 ) {
-	select {
-	case <-stop:
-		log.Info("Envoy Admin rpc stream stopped")
-		if err := stream.CloseSend(); err != nil {
-			log.Error(err, "CloseSend returned an error")
-		}
-	case err := <-processingErrorsCh:
-		if status.Code(err) == codes.Unimplemented {
-			log.Error(err, "Envoy Admin rpc stream failed, because Global CP does not implement this rpc. Upgrade Global CP.")
-			// backwards compatibility. Do not rethrow error, so KDS multiplex can still operate.
-			return
-		}
-		log.Error(err, "Envoy Admin rpc stream failed prematurely, will restart in background")
-		if err := stream.CloseSend(); err != nil {
-			log.Error(err, "CloseSend returned an error")
-		}
-		errorCh <- err
+	err := <-processingErrorsCh
+	if status.Code(err) == codes.Unimplemented {
+		log.Error(err, "rpc stream failed, because remote CP does not implement this rpc. Upgrade remote CP.")
```

Comment:
nitpick: since https://github.com//kumahq/kuma/pull/2125 we don't use "remote CP"

```suggestion
		log.Error(err, "rpc stream failed, because zone CP does not implement this rpc. Upgrade zone CP.")
```

Reply:
It's actually global cp